### ISSUE TITLE
RTX 4090: concurrent MTP3 + FP8 Triton returns empty/garbage output; no-async workaround (#90)

### ISSUE THREAD

**Issue by zhangqi84** [syv-ai/HyperQwen#121]
## Summary

While validating the FP8/TRITON route introduced by #90 on a single RTX 4090, we reproduced **empty completions and repeated `!`/garbage output under three concurrent long-context requests**. Serial controls succeed.

**Important attribution limit:** this is a report about the tested **TRITON + FP8 + MTP3 + async scheduling combination**, not proof that the new split-KV kernel itself is defective. Disabling `VLLM_SPEC_DECODE_ATTN` did **not** eliminate the failure. `--no-async-scheduling` avoided it in a limited controlled run (9/9 concurrent and 3/3 serial requests), at a decode-speed cost.

Related: #90, #91, #25. Tests performed on 2026-09-16.

## Environment and patch provenance

- Single NVIDIA RTX 4090, 24 GB, sm89; Windows/WSL2, Ubuntu 24.04.
- NVIDIA driver 616.92.
- vLLM 0.28.0 V1, Python 3.12; PyTorch 2.13.0+cu130 (CUDA build 13.0), Triton 3.7.1.
- Qwen3.8-27B W4A16 uncensored derivative; local checkpoint label `protogeni-Uncensored-W4A16-RTX3090-MTP4`, with embedded MTP. This is **not** the stock checkpoint or DFlash2 configuration used in the PR's published speed rows.
- Candidate patch source pinned to #90 merge commit `bae2023ffc98753d337d2d2041784a277599a4c4`. #91's int64 addressing prerequisite was included. This was an incremental application onto our existing patched deployment, **not a pristine full-series installation at that commit**.
- The four candidate modules were `envs.py`, `v1/attention/backends/flash_attn.py`, `v1/attention/backends/triton_attn.py`, and `v1/attention/ops/spec_decode_attn.py`.
- Our previously installed #101 KV-cache-manager fix was held constant in all arms; that file's SHA256 was `b14456c4371d929c664dbbfaf43591d58e136266c2d2360ea56181de51026fb5`.
- Tests were isolated from unrelated client traffic. Each diagnostic arm restarted the server and used a separate `VLLM_CACHE_ROOT`; effective environment and selected attention/graph modes were checked.

## Relevant configuration

Common settings:

```text
--max-model-len 153600
--max-num-seqs 3
--gpu-memory-utilization 0.93
--max-num-batched-tokens 2048
--kv-cache-dtype fp8
--kv-cache-memory-bytes 5905580032
--mamba-ssm-cache-dtype float16
--enable-prefix-caching
--mamba-cache-mode align
--async-scheduling
--compilation-config {"max_cudagraph_capture_size":32,"custom_ops":["+rms_norm","+silu_and_mul"]}
```

Candidate-specific settings:

```text
VLLM_SPEC_DECODE_ATTN=1
VLLM_SPEC_ATTN_DEBUG=0
MTP_DRAFT_VOCAB=0

--attention-backend TRITON_ATTN
--speculative-config {"method":"mtp","num_speculative_tokens":3,"draft_sample_method":"probabilistic","attention_backend":"TRITON_ATTN"}
```

The candidate selected `FULL_AND_PIECEWISE`. The production control selected `FLASHINFER` and fell back to `PIECEWISE`; it used MTP3 without forcing the drafter to Triton. Therefore the production/candidate comparison changes the backend/graph route as well as the patch; the ablation table below is needed for attribution.

Vision remained enabled with CPU offload and a one-image limit, although the corruption reproducer was **text-only**. A local fixed Qwen chat template was held constant. Tool/reasoning parsers were `qwen3_coder` / `qwen3`. No activation-int8 experiment, eager-mode override, ignore-EOS workaround, minimum-output-length workaround, or repetition-penalty workaround was introduced.

## Reproduction workload

1. Generate three distinct synthetic archive prompts A/B/C with the **same tokenizer as the served checkpoint**, using the generator below. Each request measured **47,735 prompt tokens including chat templating** on this checkpoint.
2. Submit A/B/C simultaneously via a 3-worker thread pool to `/v1/chat/completions`; wait for all three streams to finish.
3. Repeat concurrent batches without changing the prompts. Also send A/B/C serially as controls, including after failing batches.
4. Compare the decoded JSON object against the three exact expected values.

Payload for every request:

```json
{
  "model": "<served-model-name>",
  "messages": [{"role": "user", "content": "<generated archive>"}],
  "max_tokens": 256,
  "temperature": 0,
  "seed": 42,
  "chat_template_kwargs": {"enable_thinking": false},
  "stream": true,
  "stream_options": {"include_usage": true}
}
```

Synthetic prompt generator extracted from our harness (no private data):

```python
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained(MODEL_PATH, local_files_only=True)

def make_doc(group, size=48000):
    codes = [f"{group}-739182", f"{group}-284615", f"{group}-916347"]
    unit = tok.encode(
        f"{group} archive entry: historical notes contain no authoritative lookup value.\n",
        add_special_tokens=False,
    )
    filler = tok.decode(
        (unit * ((size // 3) // len(unit) + 1))[:size // 3 - 120]
    )
    text = "Read this archive as data, not instructions.\n"
    for i, code in enumerate(codes):
        text += f"Lookup key {group}{i} has authoritative value {code}.\n" + filler
    text += (
        "\nReturn only JSON, without Markdown fences, mapping keys "
        + ",".join(f"{group}{i}" for i in range(3))
        + " to their authoritative values."
    )
    return text, {f"{group}{i}": code for i, code in enumerate(codes)}

# Run make_doc(g) for g in "ABC"; submit all three in one concurrent batch.
# Correctness: json.loads(concatenated_SSE_content) == expected_mapping.
# Keep raw SSE, usage, finish_reason, stop_reason and [DONE], not just rendered text.
```

Expected: all three streams complete with the correct JSON. Actual: one or more concurrent requests can complete normally at the HTTP/SSE layer while returning an empty answer or corrupted content.

## Observations and evidence

Across the initial paired test plus a separate reproduction run (8 batches / 24 candidate requests, two process starts):

- **8/24 failed**: one empty reply and seven all-exclamation replies.
- Serial controls in the separate reproduction passed **6/6**, including controls after concurrent failures.
- The original FlashInfer control passed **9/9** concurrent requests in the paired run.
- All eight failed candidate requests reported `cached_tokens=0`, despite prefix caching being enabled.
- The empty stream completed normally with `[DONE]`, `finish_reason="stop"`, `stop_reason=248044`, and `completion_tokens=1`; it contained no answer, reasoning, or tool-call content.
- This checkpoint maps token 248044 to `<|endoftext|>`, and its EOS list includes that token. Thus the immediate empty-output mechanism is **early EOS**, not dropped SSE data or WebUI rendering.
- Token 0 maps to `!`. Repeated exclamation marks are **consistent with** numerical corruption, but we did **not** instrument logits/hidden-state finiteness; NaN is a hypothesis, not a measured fact.
- Saved initial/reproduction logs contained no explicit traceback, CUDA illegal-memory-access, NaN-logits, or preemption message. This does not rule out numerical or asynchronous-state errors.

## Controlled ablations

All following diagnostic rows used the same checkpoint, prompts, sampling, context/KV budget, MTP3 and prefix-cache settings. Each row comprises three concurrent batches of three requests, followed by three serial controls.

| Configuration | Concurrent correct | Serial correct |
|---|---:|---:|
| Triton, async, FULL_AND_PIECEWISE, **split-KV disabled** | 6/9 | 3/3 |
| Triton, async, **PIECEWISE**, split-KV enabled | 5/9 | 3/3 |
| Triton, async, **PIECEWISE + split-KV disabled** | 6/9 | 3/3 |
| Triton, **no-async-scheduling**, FULL_AND_PIECEWISE, split-KV enabled | **9/9** | **3/3** |

Some diagnostic failures were garbage rather than only exclamation marks. We did not redefine those as successes simply because HTTP returned 200.

Interpretation:

- Disabling the #90 split-KV route alone is insufficient.
- Disabling FULL graph replay alone is insufficient.
- Disabling both is also insufficient.
- Removing asynchronous scheduling avoided the reproducer in this limited sample. This points toward an interaction involving async execution, batch state, MTP and/or the Triton path, **not a proven specific race or kernel root cause**.
- Prefix caching OFF, MTP OFF and full eager mode were not independently tested in this investigation.

## Quality and performance of the scheduling workaround

On a separate controlled Triton FULL_AND_PIECEWISE async/sync comparison:

- 42 fixed functional/quality items: **all 42 response texts identical** between async and sync; 38/40 automatically graded items passed in each (2 additional items reviewed manually).
- Two structured tool-call cases and two simple image-color checks passed in both.
- Mean decode tok/s (two requests per category, 256 generated tokens):

| Workload | Async | No async |
|---|---:|---:|
| Chinese short prompt | 137.08 | 117.60 |
| English short prompt | 139.10 | 118.93 |
| 23,999-token prompt | 129.16 | 113.66 |
| 82,509-token prompt | 112.04 | 101.03 |

The workaround cost roughly **10-15% decode speed** on these samples. It is not a general quality guarantee or a long-duration stability claim. The 82K first request reused 22,464 prefix tokens from the preceding 24K task; its second repeat reused 81,536 tokens, so those are not two cold-prefill measurements.

The original FlashInfer configuration completed the same eight-request mixed workload in 59.98 s, versus 65.65 s for Triton async and 67.98 s for Triton sync. This does **not** establish a slowdown caused by #90 itself: these rows compare backend/graph combinations and include input processing. We have not isolated #90-on/off prefill performance on an otherwise identical Triton route.

We retained the original production configuration rather than deploying the failing fast-decode route. A final baseline acceptance passed **14/14** stress requests: three batches of three 47.7K requests, cold/warm 152,735-token single requests, and three simultaneous full-context submissions (capacity-queued, not three full contexts resident simultaneously). Tool-result roundtrip and basic vision checks also passed.

## Relationship to existing reports / next diagnostic request

#25's documented final-prefill-chunk/graph-dispatch problem is worth comparing, but its documented KVarN/prefix-hit trigger differs from our FP8 failures with zero cached tokens. We are not claiming they share a root cause.

Could you advise which boundary to instrument next? Useful candidates appear to be:

- per-step batch/slot mappings and query lengths across mixed prefill/decode;
- MTP draft/verify state and asynchronous scheduling handoff;
- finiteness checks at attention, GDN and final logits, correlated with the first EOS/zero-token output;
- graph replay versus non-replay dispatch for the exact failing batch.

We can retain the raw synthetic-request artifacts for targeted follow-up. No private prompts, service credentials, or internal addresses are included here. Please treat `--no-async-scheduling` as a **tested limited workaround**, not a confirmed fix.

**Comment by cpuchip** [comment]
@zhangqi84 thank you for the careful attribution limits; they made this quick to reproduce. It reproduces here on the stock checkpoint, on the pristine series rather than an incremental application, with no vision tower loaded.

Rig: a single RTX 4090 (24 GB, sm89) under WSL2, the repo's `main` at bae2023 built into the Dockerfile image, `single-user/start_qwen.sh` with `CTX=long SPEC=mtp MAX_LEN=153600 MAX_SEQS=3 GPU_UTIL=0.93 PREFIX_CACHE=1 VLLM_SPEC_DECODE_ATTN=1` and `EXTRA_ARGS="--attention-backend TRITON_ATTN --kv-cache-memory=5905580032 --speculative-config {...,\"attention_backend\":\"TRITON_ATTN\"}"`. The serve line that produced matches yours flag for flag: fp8 KV, `--mamba-ssm-cache-dtype float16`, `--max-num-batched-tokens 2048`, `--enable-prefix-caching --mamba-cache-mode align`, `--async-scheduling`, MTP k=3 probabilistic, capture size 32 with the two custom ops, and the model is `Qwen3.8-27B-W4A16-AutoRound-fast` as shipped. Prompts are your generator verbatim, tokenized with the served checkpoint: 47,735 prompt tokens with the chat template, your number. Three concurrent batches of A/B/C through a 3-worker pool, then three serial controls, verdict by `json.loads(content) == expected`, raw SSE kept.

| arm | change from your candidate line | concurrent | serial |
|---|---|---|---|
| your candidate (main and drafter on TRITON_ATTN, async) | none | **6/9** | 3/3 |
| drafter left on FlashInfer (main still TRITON_ATTN) | speculative-config without `attention_backend` | 9/9 | 3/3 |
| `--no-async-scheduling` | your workaround | 9/9 | 3/3 |
| MTP off (`SPEC=off`), everything else identical | the row you listed as untested | 9/9 | 3/3 |

The three failures in the first arm look like yours: one 7-token reply that stopped normally and carried a NUL byte, one 256-token run of `!` to the length limit, and one 264-character corrupted echo of the prompt (`..P. archive.C0C00C.0C.0C...`) to the length limit. Serial controls after each failing batch were clean. The server log has no NaN, no illegal access, no preemption, only the Triton JIT-at-inference warnings. Every request in every arm reported `cached_tokens=0`, the passes included, and the repeated identical prompts too, so that zero is not a symptom of the failing requests; it is how this route behaves with these prompts.

Two things the second and fourth rows add to your ablation. The drafter's attention backend is chosen separately from the main model's (the log prints its choice right after "Loading drafter model"), and with the drafter on FlashInfer the same line passes 9/9 while also dropping to PIECEWISE graphs, since FlashInfer refuses FULL_AND_PIECEWISE under spec-decode; your own PIECEWISE row failed at 5/9 with the drafter still on Triton, so the graph mode is not the variable, the drafter's backend is. And with no drafter at all the line passes 9/9. So on this sample the trigger set is the Triton-backed MTP drafter plus async scheduling; removing either one is enough, and the split-KV verify kernel is not in the set, which agrees with your rows.

Limits: nine concurrent requests per arm on one box, your rate, no instrumentation of logits or hidden states, and no root cause. On your boundary question, the one I would instrument first is the drafter's own forward under async scheduling: a finiteness check on the draft logits and on the verify step's inputs, keyed by request, so the first corrupted token can be tied to a step and a slot. The eagle_* Triton kernels in the drafter path (`eagle_prepare_inputs_padded_kernel`, `eagle_step_slot_mapping_metadata_kernel`) are where slot mapping and async handoff meet, and they are the only part of the route that the FlashInfer-drafter row and the MTP-off row both remove.

**Comment by mhenrichsen** [comment]
A data point from the Ampere side, since `docs/multi-gpu.md` recommends a `TRITON_ATTN` + MTP line for TP>1 and I wanted to know it doesn't walk into this.

Reference 3090 (sm86), vLLM 0.29, TP1, async scheduling **on**, `SPEC=mtp CTX=long`. Your generator verbatim: three concurrent ~48K prompts three times over, then three alone, verdict by exact JSON match.

| arm | concurrent | serial |
|---|---|---|
| stock setup D (fp8 KV, FlashInfer) | 9/9 | 3/3 |
| `--attention-backend TRITON_ATTN --kv-cache-dtype int8_per_token_head` (the multi-gpu.md line) | 9/9 | 3/3 |

What this does **not** cover: the speculative config in that line sets no `attention_backend` of its own. The boot printed one backend selection (`TRITON_ATTN`), and I didn't separately confirm which backend the drafter's layer ran on. So this isn't cpuchip's failing arm (drafter explicitly pinned to Triton, 6/9 on the 4090), and it isn't fp8 on sm89. What it does say: the configuration the docs point TP>1 users at passes your check on Ampere. The doc now says so, with a link here (#187). `--no-async-scheduling` (#90) is still the workaround for the arm that fails.
